app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java

```diff
@@ -281,61 +367,80 @@ private void startApp() {
+        mainIntent.putExtra(MainActivity.EXTRA_RESET_DATABASE, mResetDatabase);
+        startActivity(mainIntent);
+        finish();
+    }
 
-        /**
-         * In case start_into_web_gui option is enabled, start both activities
-         * so that back navigation works as expected.
-         */
-        if (mPreferences.getBoolean(Constants.PREF_START_INTO_WEB_GUI, false)) {
-            startActivities(new Intent[] {mainIntent, new Intent(this, WebGuiActivity.class)});
-        } else {
-            startActivity(mainIntent);
+    private boolean hasLocationPermission() {
+        for (String perm : PermissionUtil.getLocationPermissions()) {
+            if (ContextCompat.checkSelfPermission(this, perm) != PackageManager.PERMISSION_GRANTED) {
+                return false;
+            }
         }
-        finish();
+        return true;
     }
 
     /**
      * Permission check and request functions
      */
     private void requestLocationPermission() {
         ActivityCompat.requestPermissions(this,
-                Constants.getLocationPermissions(),
+                PermissionUtil.getLocationPermissions(),
                 Constants.PermissionRequestType.LOCATION.ordinal());
     }
 
-    private boolean haveStoragePermission() {
-        int permissionState = ContextCompat.checkSelfPermission(this,
-                Manifest.permission.WRITE_EXTERNAL_STORAGE);
-        return permissionState == PackageManager.PERMISSION_GRANTED;
+    private void requestStoragePermission() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+            requestAllFilesAccessPermission();
+        } else {
+            ActivityCompat.requestPermissions(this,
+                    new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
+                    Constants.PermissionRequestType.STORAGE.ordinal());
+        }
     }
 
-    private void requestStoragePermission() {
-        ActivityCompat.requestPermissions(this,
-                new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
-                Constants.PermissionRequestType.STORAGE.ordinal());
+    @TargetApi(30)
+    private void requestAllFilesAccessPermission() {
+        Intent intent = new Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION);
+        intent.setData(Uri.parse("package:" + getPackageName()));
+        try {
+            ComponentName componentName = intent.resolveActivity(getPackageManager());
+            if (componentName != null) {
+                // Launch "Allow all files access?" dialog.
+                startActivity(intent);
+                return;
+            }
+            Log.w(TAG, "Request all files access not supported");
+        } catch (ActivityNotFoundException e) {
+            Log.w(TAG, "Request all files access not supported", e);
+        }
+        Toast.makeText(this, R.string.dialog_all_files_access_not_supported, Toast.LENGTH_LONG).show();
```

Comment:
Might be interesting which branch we've hit?

Also, I guess the user can just press cancel or something, but I guess that wouldn't allow them to skip to the next slide, as it checks.

Reply:
It's already discernable in the logs - I wouldn't display any such low-level info in a toast. And generally it's mainly a very defensive measure, not likely that either case is hit.

And yes it does check, display a message and prevent continuation, if the user tries to continue without having approved the access.